src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+        // [x] If Queue doesn't support GRAPHICS_BIT or COMPUTE_BIT ->  must be multiple of 4
+        // [x] bufferOffset must be a multiple of texel block size in bytes
+    bufferOffsetAlignment = dstImageTexelBlockInfo.getBlockByteSize();
+    if(asset::isDepthOrStencilFormat(dstImageFormat))
+        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
+
+    bool queueSupportsCompute = queueFamilyProps.queueFlags.hasFlags(IPhysicalDevice::EQF_COMPUTE_BIT);
+    bool queueSupportsGraphics = queueFamilyProps.queueFlags.hasFlags(IPhysicalDevice::EQF_GRAPHICS_BIT);
+    if((queueSupportsGraphics || queueSupportsCompute) == false)
+        bufferOffsetAlignment = std::lcm(bufferOffsetAlignment, 4u);
+    // TODO: Need to have a function to get equivalent format of the specific plane of this format (in aspectMask)
+    // if(asset::isPlanarFormat(dstImageFormat->getCreationParameters().format))
+    assert(core::is_alignment(bufferOffsetAlignment));
+        
+    // Queues supporting graphics and/or compute operations must report (1,1,1) in minImageTransferGranularity, meaning that there are no additional restrictions on the granularity of image transfer operations for these queues.
+    // Other queues supporting image transfer operations are only required to support whole mip level transfers, thus minImageTransferGranularity for queues belonging to such queue families may be (0,0,0)
+    canTransferMipLevelsPartially = !(minImageTransferGranularity.width == 0 && minImageTransferGranularity.height == 0 && minImageTransferGranularity.depth == 0);
+}
+
+size_t ImageRegionIterator::getMemoryNeededForRemainingRegions() const
+{
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+    assert(dstImageTexelBlockInfo.getBlockByteSize()>0u);
+    auto texelBlockDim = dstImageTexelBlockInfo.getDimension();
+    size_t memoryNeededForRemainingRegions = 0ull;
+    for (uint32_t i = currentRegion; i < regions.size(); ++i)
+    {
+        memoryNeededForRemainingRegions = core::alignUp(memoryNeededForRemainingRegions, bufferOffsetAlignment);
+
+        const asset::IImage::SBufferCopy & region = regions[i];
+
+        auto subresourceSize = dstImage->getMipSize(region.imageSubresource.mipLevel);
+
+        assert(static_cast<uint32_t>(region.imageSubresource.aspectMask) != 0u);
+        assert(core::isPoT(static_cast<uint32_t>(region.imageSubresource.aspectMask)) && "region.aspectMask should only have a single bit set.");
+        // Validate Region
+        // canTransferMipLevelsPartially
+        if(!canTransferMipLevelsPartially)
+        {
+            assert(region.imageOffset.x == 0 && region.imageOffset.y == 0 && region.imageOffset.z == 0);
+            assert(region.imageExtent.width == subresourceSize.x && region.imageExtent.height == subresourceSize.y && region.imageExtent.depth == subresourceSize.z);
+        }
```

Comment:
you shouldn't assert these, you shouldn't have been able to construct an iterator in the first place with invalid regions due to:
- whole mip requirement
- offset and extent not being multiples of granularity or texel block dimensions
etc.